### ISSUE TITLE
Have CUSTOM_ID be available in canvas_item vertex shader

### ISSUE THREAD

**Issue by Hasenn** [godotengine/godot-proposals#1525]
**Describe the project you are working on:**
It's a 2D platformer
**Describe the problem or limitation you are having in your project:**
We're trying to implement a "flipbook" variation on a particle effect (having a random texture on each instance of some 2D GPU particles) and we concluded we either have to rewrite the whole default particle shader just to pass some random seed through INSTANCE_CUSTOM, or resort to some hacks using the hue variation in the default process material and later read it in the canvas_item material.
**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
have CUSTOM_ID also be a thing in canvas_item, i would be able to hash that and add some variation to my particle's rendering based on that
**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
```glsl
varying float lifetime;
varying float rand;
void vertex(){
        lifetime = INSTANCE_CUSTOM.y; //currently this works
        rand = rand_from_seed(hash(INSTANCE_ID)); //this doesn't in 2D shaders
    }
//then use rand to influence rendering in the fragment function
```
**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
The workaround is not pretty but it can be done in about 10 lines of shader code, and we also have to move textures from where they usually go to inside the shader, and use the regular texture field with something like a green square which is unintuitive for our designers
**Is there a reason why this should be core and not an add-on in the asset library?:**
it already is a thing for 3D particles, i don't see why it shouldn't also work for 2D

**Comment by Calinou** [comment]
Shouldn't we just implement flipbook animation in the default 2D particle shaders if it's already in 3D?

**Comment by Hasenn** [comment]
I meant that INSTANCE_ID was a thing in 3D, i don't know about any flipbook animation in 3D

**Comment by Hasenn** [comment]
What i mean by "flipbook" is an idea from riot games https://na.leagueoflegends.com/en-us/news/dev/dev-a-storm-is-brewing/
you can find it in that link by CTRL-Fing "particle"

**Comment by Calinou** [comment]
@Hasenn I know what particle flipbooks are about :slightly_smiling_face: 

They're already supported in both 2D and 3D. In 2D, you need to create and assign a CanvasItemMaterial with the [particles animation mode](https://docs.godotengine.org/en/latest/classes/class_canvasitemmaterial.html#class-canvasitemmaterial-property-particles-animation) configured.

See this project: [particle-flipbook-2d.zip](https://github.com/godotengine/godot-proposals/files/5241600/particle-flipbook-2d.zip)

**Comment by Hasenn** [comment]
hmm this is cool but we are already using a ShaderMaterial (for some alpha erosion) in the CanvasItemMaterial's slot, is there a way to extend it with our shader or use both ? or to replicate what it does in our custom canvas_item shader ?

**Comment by clayjohn** [comment]
@Hasenn certainly! Just create a new CanvasItemMaterial, enable particle_animation and then select the material and "convert to shader material" then it will give you all the code and you can just copy it into your own shader.

**Comment by Hasenn** [comment]
ooh i didn't know that was a thing that's so cool thanks
i don't know if i should close this as i can still imagine `CUSTOM_ID` being usefull for other use cases

**Comment by clayjohn** [comment]
I think it is worth leaving up for now. We will see if others have unique use cases.

**Comment by Sebjugate** [comment]
If I'm getting the gist of this request correctly, it is for `INSTANCE_ID` to be available in canvas_item shaders like it is in 3D for particles and multimesh. If that is the case, I'd also like to see this built-in be available.

I think the `CUSTOM_ID` name was maybe a typo?

**Comment by Hasenn** [comment]
@Sebjugate Yeah i think i meant `INSTANCE_ID`

**Comment by clayjohn** [comment]
Implemented in https://github.com/godotengine/godot/pull/54791!

This should already be available in all the alphas creates so far :)
